[WIP]PR#11 ( renderFormFeaturedImageAndCaption )

rgadon107 opened the pull request:
## PR Summary 

This pull request includes unit and integration tests for the function `render_form_featured_image_and_caption()` in the `extend-give-wp` plugin. The relative file path to the function under test is: 

>`extend-give-wp/src/support/load-assets.php`.

### Issues -- Unit Test

The unit and integration tests for this function are complicated by the fact that there are 2 variables, and a WP core function passed into the view. To simplify the testing scenarios, I removed the function `wp_get_attachment_image()` from the expected view in both the unit and test classes. That function returns an enormous `<img>` tag element that I currently find too complicated to mock or match with an actual value. 

I chose instead to test the HTML at the start and end of the view. Each HTML segment contains a variable escaped with `esc_attr()`.  [ I am now able to stub out `esc_attr()`. Thanks for that hint earlier.]

I can pass the `$post_excerpt` associated with the featured image into the view. I cannot pass the featured image ID ( `$options['featured-image-id']` or ‘$attachment_id’ ) into the actual view. The source code checks that `$options['featured-image-id’]` is set. If true, it returns the database value and assigns it to `$attachment_id`. 

In the unit test, I removed the empty data set, and ran a test assertion to check that the option value for 'featured-image-id' was set. It was ( returned true ). Yet, the test method processes the ternary operator and returns falsey ( 0 ). 

Resolving this issue is important. It will occur again when I attempt to write a unit test for the function `render_featured_image_id_field()` in the file: 

>`extend-give-wp/src/admin/option-settings-admin.php`. 

![renderFormFeaturedImage_unit_test_error_message_05-14-20](https://user-images.githubusercontent.com/3718603/81991192-43d19400-9606-11ea-86a3-61fcd5d9c655.png)

### Issues -- Integration Test

The most recent commits to the integration test file employed a different strategy. I used `add_option()` to add the option `extend-give-wp` to the database, and then used `get_option( `extend-give-wp`, [] )` to retrieve it. I then ran 2 test assertions to check that the both the option_name and the key=>value pair added to the option were set in memory ( they were ). That told me I could expect that the variable `$attachment_id` would be truthy and return the expected value ( it did ). 

My challenge is how to set or update a value on `$attachment[ 'post_excerpt' ]`. At present, the variable `$post_excerpt` returns falsey ( empty string ). I need to update the attachment object via the WordPress factory() method. I've attempted a couple of Core functions and methods and passing in a set of $args: 

- `wp_update_post()`, 
- `wp_insert_attachment()`,
- `$this->factory()->attachment->create_object( $args )`. 

Each fails to return an updated attachment or attachment object ( they return a value of 0 ). 
The result is that the variable `$post_excerpt` fails to load in the actual view file. 

![renderFormFeaturedImage_integration_test_error_messages_05-15-20](https://user-images.githubusercontent.com/3718603/82102491-5cf54600-96d5-11ea-9b5e-23e0f22792e3.png)

rgadon107 committed:
[WIP]Commit unit test file. 

- Test assertion does not run. Stub of core function 'esc_attr' in 
TestCase is not mocked.

rgadon107 committed:
PHPCS fixes.

rgadon107 committed:
Disable phpcs for missing paramater tag in test method docblock.

rgadon107 committed:
Merge branch 'master' into renderFormFeaturedImageAndCaption

rgadon107 committed:
Get the 'post_excerpt' via the Core function 'get_post_field'.

rgadon107 committed:
[WIP]Commit integration test file.

- $post_excerpt is falsey (empty string). 
- Use a factory to set an $attachment_id and $post_excerpt. 
- You're also going to need to add (set) the wp_attachment_image data to 
the database in order to call (get) it.

rgadon107 committed:
Update integration test file. 

- Add dataProvider params to test method ( $args, $starts_with, $ends_with ).
- Use WP factories for post and attachment to generate IDs.
- Add test assertions to test that the view file starts and ends with a specific string.
- Update dataProvider.

rgadon107 committed:
Mock function 'get_post_field'.

rgadon107 committed:
Fix function mocks. 

- Update params passed to 'get_post_field'. 
- Mock 'esc_attr' with when()->justEcho().

hellofromtonya commented on tests/phpunit/Unit/support/renderFormFeaturedImageAndCaption.php:
```diff
@@ -0,0 +1,97 @@
+			->with( $post_data['attachment_id'], 'large', false, [ 'class' => 'featured-image' ] )
+			->andReturn( $html );
+		Functions\expect( '_get_plugin_dir' )->andReturn( EXTEND_GIVE_WP_ROOT_DIR );
+		Functions\when( 'esc_attr')->justEcho();
```
`esc_attr` does not echo, it returns the first argument after cleaning it. You can stub it and others by adding this code to your test case https://github.com/wp-media/phpunit/blob/master/Unit/TestCase.php#L63

rgadon107 replied:
@hellofromtonya Per your comment on PR#16, I added `$this->setup_common_wp_stubs()` to the `setUp()` method in the unit test class. That resolved the problem with `esc_attr()` being undefined.

rgadon107 committed:
Merge branch 'master' into renderFormFeaturedImageAndCaption

rgadon107 committed:
Update unit test file.

- Change dataProvider params in test method.
- Change function mocks and param passed to function under test.
- Include 2 data sets in dataProvider; 'post data is empty', & 'post data is non-empty'. 
- Remove expected HTML for core function 'wp_get_attachment_image' due to difficuly in returning an actual value from the view.

rgadon107 committed:
Unit test file updates.

- Add TestCase method 'setup_commong_wp_stubs' to class setUp method.
- Change dataProvider params as inputs to unit test method.
- Change function mocks. 
- Update datasets to test when 'post data is empty' and 'post data is valid.

rgadon107 committed:
Unit test file updates.

- Add data set to test scenario that $form_id and $featured_image_id ($attachment_id) are empty.

rgadon107 committed:
Add integration file updates.

- Add data set to test scenario that $form_id, $post_excerpt, views are empty. 
- Note: test throws 2 errors: (1) expected value of view in first data set cannot be empty; (2) $form_id in the second data set must be a string. Array given.

rgadon107 committed:
Change the return value of 'get_option' and input to 'get_post_field. 

- Doesn't change test result.

rgadon107 committed:
Change string literal key in the $options array to match source code.

rgadon107 committed:
Integration test file updates

- Add 'tearDown()' method to test class to delete option after each run of the test method. 
- Correct the priority number and qualified name of the function under test in the first test method.
- Add the post factory back into the test method. 
- Add and get the option 'extend-give-wp'. 
- Test assertions that the option was set and added to the database. Var $attachment_id now returns expected value. 
- Create and update an attachment fixture for the test method. 
- Notes included on unsuccessful attempts to update 'post_excerpt' of the attachment object. Needed to render 'post_except in the view. 
- Commented out test assertions in test method to reduce the level of 'noise' returned when running test method. 
- Deleted empty data set in dataProvider to simplify output of test method.
